tools/clang/lib/SPIRV/SpirvBuilder.cpp

```diff
@@ -86,5 +87,576 @@ void SpirvBuilder::setContinueTarget(SpirvBasicBlock *continueLabel) {
+    mask = mask | ImageOperandsMask::Lod;
+  }
+  if (grad.first && grad.second) {
+    mask = mask | ImageOperandsMask::Grad;
+  }
+  if (constOffset) {
+    mask = mask | ImageOperandsMask::ConstOffset;
+  }
+  if (varOffset) {
+    mask = mask | ImageOperandsMask::Offset;
+    requireCapability(spv::Capability::ImageGatherExtended);
+  }
+  if (constOffsets) {
+    mask = mask | ImageOperandsMask::ConstOffsets;
+    requireCapability(spv::Capability::ImageGatherExtended);
+  }
+  if (sample) {
+    mask = mask | ImageOperandsMask::Sample;
+  }
+  if (minLod) {
+    requireCapability(spv::Capability::MinLod);
+    mask = mask | ImageOperandsMask::MinLod;
+  }
+  return mask;
+}
+
+SpirvInstruction *SpirvBuilder::createImageSample(
+    QualType texelType, QualType imageType, SpirvInstruction *image,
+    SpirvInstruction *sampler, bool isNonUniform, SpirvInstruction *coordinate,
+    SpirvInstruction *compareVal, SpirvInstruction *bias, SpirvInstruction *lod,
+    std::pair<SpirvInstruction *, SpirvInstruction *> grad,
+    SpirvInstruction *constOffset, SpirvInstruction *varOffset,
+    SpirvInstruction *constOffsets, SpirvInstruction *sample,
+    SpirvInstruction *minLod, SpirvInstruction *residencyCode,
+    SourceLocation loc) {
+  assert(insertPoint && "null insert point");
+
+  // The Lod and Grad image operands requires explicit-lod instructions.
+  // Otherwise we use implicit-lod instructions.
+  const bool isExplicit = lod || (grad.first && grad.second);
+  const bool isSparse = (residencyCode != nullptr);
+
+  spv::Op op = spv::Op::Max;
+  if (compareVal) {
+    op = isExplicit ? (isSparse ? spv::Op::OpImageSparseSampleDrefExplicitLod
+                                : spv::Op::OpImageSampleDrefExplicitLod)
+                    : (isSparse ? spv::Op::OpImageSparseSampleDrefImplicitLod
+                                : spv::Op::OpImageSampleDrefImplicitLod);
+  } else {
+    op = isExplicit ? (isSparse ? spv::Op::OpImageSparseSampleExplicitLod
+                                : spv::Op::OpImageSampleExplicitLod)
+                    : (isSparse ? spv::Op::OpImageSparseSampleImplicitLod
+                                : spv::Op::OpImageSampleImplicitLod);
+  }
+
+  // minLod is only valid with Implicit instructions and Grad instructions.
+  // This means that we cannot have Lod and minLod together because Lod requires
+  // explicit insturctions. So either lod or minLod or both must be zero.
+  assert(lod == nullptr || minLod == nullptr);
+
+  if (isSparse) {
+    requireCapability(spv::Capability::SparseResidency);
+  }
+
+  // An OpSampledImage is required to do the image sampling.
+  auto *sampledImage =
+      new (context) SpirvSampledImage(imageType, /*id*/ 0, loc, image, sampler);
+  insertPoint->addInstruction(sampledImage);
+
+  if (isNonUniform) {
+    // The sampled image will be used to access resource's memory, so we need
+    // to decorate it with NonUniformEXT.
+    // TODO: decorateNonUniformEXT(sampledImgId);
```

Comment:
This will be re-enabled in my next commit. (when decorate* methods are implemented)